File: Algorithms/0684_Redundant_Connection/Python/Redundant_Connection_Solution_1.py

# Space: O(n)
# Time: O(n)

import collections

class Solution:
    def findRedundantConnection(self, edges):
        cache = collections.defaultdict(set)
        res = None

        def dfs(start, end, visited):
            if start == end: return True
            if start in visited: return False
            visited.add(start)

            for i in cache[start]:
                if dfs(i, end, visited): return True
            return False

        for i, j in edges:
            if dfs(i, j, set()):
                res = [i, j]
            else:
                cache[i].add(j)
                cache[j].add(i)

        return res

File: Algorithms/0684_Redundant_Connection/Java/RedundantConnectionSolution1.java

// Space: O(n)
// Time: O(n)

class Solution {
    private Map<Integer, Set<Integer>> graph = new HashMap<>();

    public int[] findRedundantConnection(int[][] edges) {
        int[] res = null;
        for (int[] edge : edges) {
            if (dfs(edge[0], edge[1], new HashSet<>())) res = edge;

            Set<Integer> val = graph.getOrDefault(edge[0], new HashSet<>());
            val.add(edge[1]);
            graph.put(edge[0], val);

            Set<Integer> val1 = graph.getOrDefault(edge[1], new HashSet<>());
            val1.add(edge[0]);
            graph.put(edge[1], val1);
        }
        return res;
    }

    private boolean dfs(int start, int end, Set<Integer> visited) {
        if (visited.contains(start)) return false;
        if (start == end) return true;
        visited.add(start);

        for (Map.Entry<Integer, Set<Integer>> entry : graph.entrySet()) {
            for (Integer i : entry.getValue()) {
                if (dfs(i,end,visited)) return true;
            }
        }
        return false;
    }
}




